File: backend/src/api/schemas/runs.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


Provider = Literal["aws", "azure"]
ProviderScope = Literal["aws", "azure", "both"]


class CreateRunRequest(BaseModel):
    query: str = Field(min_length=1)
    provider_scope: ProviderScope
    constraints: dict[str, Any] = Field(default_factory=dict)
    optimization_preferences: dict[str, Any] = Field(default_factory=dict)


class CreateRunResponse(BaseModel):
    run_id: str
    status: Literal["queued", "running"] = "running"


class EvidenceReference(BaseModel):
    source_uri: str
    title: str
    excerpt: str | None = None
    availability_status: Literal["available", "unavailable"] = "available"


class ValidationFinding(BaseModel):
    finding_id: str
    iteration: int
    domain: Literal["compute", "storage", "network", "database", "cross_domain"]
    outcome: Literal["pass", "fail", "warn"]
    severity: Literal["critical", "high", "medium", "low"]
    reason: str
    remediation: str | None = None
    evidence: list[EvidenceReference] = Field(default_factory=list)


class ProviderRunDetail(BaseModel):
    provider: Provider
    status: Literal["generating", "validating", "optimizing", "accepted", "stopped", "failed"]
    iteration_count: int
    accepted: bool
    terminal_reason: str | None = None
    unresolved_findings: int = 0
    latest_output: str | None = None
    failure_reason: str | None = None
    refinement_feedback: list[dict[str, Any]] = Field(default_factory=list)


class RunDetail(BaseModel):
    run_id: str
    query: str
    provider_scope: ProviderScope
    provider_runs: list[ProviderRunDetail]


class HistoryItem(BaseModel):
    run_id: str
    query: str
    provider_scope: ProviderScope
    created_at: datetime
    provider_runs: list[ProviderRunDetail]


class HistoryResponse(BaseModel):
    items: list[HistoryItem]


class FindingListResponse(BaseModel):
    items: list[ValidationFinding]

File: frontend/src/services/runs.ts
import { apiRequest } from "./api-client";

export type Provider = "aws" | "azure";
export type ProviderScope = Provider | "both";

export interface CreateRunRequest {
  query: string;
  provider_scope: ProviderScope;
  constraints?: Record<string, unknown>;
  optimization_preferences?: Record<string, unknown>;
}

export interface CreateRunResponse {
  run_id: string;
  status: "queued" | "running";
}

export interface ProviderRunDetail {
  provider: Provider;
  status: "generating" | "validating" | "optimizing" | "accepted" | "stopped" | "failed";
  iteration_count: number;
  accepted: boolean;
  terminal_reason?: string;
  unresolved_findings: number;
  latest_output?: string;
  failure_reason?: string;
}

export interface RunDetail {
  run_id: string;
  query: string;
  provider_scope: ProviderScope;
  provider_runs: ProviderRunDetail[];
}

export interface EvidenceReference {
  source_uri: string;
  title: string;
  excerpt?: string;
  availability_status: "available" | "unavailable";
}

export interface ValidationFinding {
  finding_id: string;
  iteration: number;
  domain: "compute" | "storage" | "network" | "database" | "cross_domain";
  outcome: "pass" | "fail" | "warn";
  severity: "critical" | "high" | "medium" | "low";
  reason: string;
  remediation?: string;
  evidence: EvidenceReference[];
}

export async function createRun(payload: CreateRunRequest): Promise<CreateRunResponse> {
  return apiRequest<CreateRunResponse>("/api/v1/runs", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function getRun(runId: string): Promise<RunDetail> {
  return apiRequest<RunDetail>(`/api/v1/runs/${runId}`);
}

export async function getHistory(provider?: Provider): Promise<{ items: RunDetail[] }> {
  const suffix = provider ? `?provider=${provider}` : "";
  return apiRequest<{ items: RunDetail[] }>(`/api/v1/history${suffix}`);
}

export async function getFindings(runId: string, provider: Provider): Promise<{ items: ValidationFinding[] }> {
  return apiRequest<{ items: ValidationFinding[] }>(`/api/v1/runs/${runId}/providers/${provider}/findings`);
}
